### ISSUE TITLE
invalid ids in the Manual html

### ISSUE THREAD

**Issue by sanette** [ocaml/v2.ocaml.org#1093]
See 
https://github.com/aantron/lambdasoup/issues/26

Apparently hevea uses the char ":" in CSS identifiers, while it's forbidden by CSS syntax.
It appears frequently in the Ocaml Manual, for instance: `<a id="c:core-xamples"></a>`
It prevents from doing some CSS selections, for instance with lambdasoup.

**Comment by Chris00** [comment]
HTML4 [allows colons in ID values](https://www.w3.org/TR/html4/types.html#type-id) and HTML5 is even more permissive.  So IMHO, it's lambdasoup that must be more updated.

**Comment by Chris00** [comment]
Now, we can see if a middle ground is possible — but the manual sources are in the same repository as the compiler https://github.com/ocaml/ocaml/tree/trunk/manual

**Comment by sanette** [comment]
so it seems that each side has a good reason for not moving... As far as I'm concerned, that's not really a problem; I can use a workaround.
 I was just mentioning this in case some party had a clear issue to correct.

**Comment by aantron** [comment]
@Chris00, In general, HTML5 is more permissive about *parsing* (in order to parse all the messy HTML out there), and more strict about *emitting* (in order to avoid creating any more such). Do you have a reference for HTML5 specifically allowing `:`?

For comparison, try this HTML file in your browser:

```html
<html>
<head>

<style>
#test:2 {
  background-color: red;
}
</style>

</head>
<body>

<p id="test:2">foo</p>

</body>
</html>
```

In my Chromium-based browser, this selector fails to match on the `p` element, though Chromium parses and interprets according to HTML5. Generally speaking, Markup.ml and Lambda Soup behave like browsers when parsing, with the same error recovery, and with only a few obscure deviations, this not being one of them to my knowledge :)

Given the above, I don't think this issue is decided according to the HTML5 spec directly (the less so HTML4), but according to CSS specs. An id with a literal `:` in it causes problems with pseudo-classes, so my obvious speculation is that `:` has been effectively disallowed for a decade or more for that reason.

**Comment by aantron** [comment]
I just glanced through the HTML5 spec again, and my impression is that:

1. HTML5 does not itself forbid `:`.
2. From the CSS spec, using `:` in an id effectively makes it impossible to select that id using CSS.

So, `:` is "allowed" by HTML5, but it's best to follow the more restrictive requirements of CSS and not emit it.

**Comment by aantron** [comment]
Looking at CSS1 ["Forward-compatible parsing"](https://www.w3.org/TR/1999/REC-CSS1-19990111#forward-compatible-parsing), it looks like literal `:` has been disallowed in CSS selectors from the beginning, 1999 at the latest, 1996 at the earliest &mdash; search that section for "in CSS1". Given HTML4 was released in the same time frame, I'd say the HTML4 spec alone has never been sufficient to decide whether it is wise to emit `:` in documents that will be interpreted with CSS. And on that, W3C has been recommending using HTML with CSS since 1997.

**Comment by Chris00** [comment]
@aantron Thanks for the detailed explanation.  [Your example](https://github.com/ocaml/ocaml.org/issues/1093#issuecomment-549702852) neither works on Firefox.  However links such as https://caml.inria.fr/pub/docs/manual-ocaml-4.09/index.html#p%3Atutorials do.  If I understand correctly, they also need to be changed because lambdasoup will not be able to handle them?

**Comment by aantron** [comment]
Yes, the explanation for links working is of course that they have nothing to do with CSS selectors. It's only CSS selectors that don't allow `:`. Lambda Soup has no problem with links, whether they contain `:` or not, but the links would change as a side effect of changing ids. The only thing Lambda Soup doesn't handle is `:` in CSS selectors (not ids, it parses ids with `:` fine, it just can't address them with CSS). This is the same behavior as in browsers.

**Comment by aantron** [comment]
i.e. the DOM with ids containing `:` is loaded fine, and links into it work, but the elements with ids with `:` are specifically not reachable by CSS id selectors (EDIT: at least not in the obvious way).

**Comment by Chris00** [comment]
@aantron Thanks.

@sanette Is this issue a blocker to [mirror the manual on o.o](https://github.com/ocaml/ocaml.org/issues/1011) because you use lambdasoup?  Does that imply that older versions of the manual will not be mirrored?

**Comment by sanette** [comment]
@Chris00, no, as I said, this is not a blocker for me. Instead of using a CSS selector, I can parse the `id` field.

**Comment by aantron** [comment]
Indeed, there should be at least two other ways to select by id:

1. Using an attribute selector (`[id=foo]`), which is what I think @sanette is using.
2. Using the functional interface with `Soup.filter`, `Soup.id`, and `Soup.first`.

Lambda Soup may also support escape sequences in id selectors, but I am not immediately sure about this.

**Comment by Chris00** [comment]
In CSS, it seems that it seems that to escape `:`, it suffices to precede it with a `\`.  The following works (in Firefox):
```html
<html>
<head>

<style>
#test\:2 {
  background-color: red;
}
</style>

</head>
<body>

<p id="test:2">foo</p>

</body>
</html>
```
Is it still necessary to modify the manual or hevea in view of this?

**Comment by dbuenzli** [comment]
> Is it still necessary to modify the manual or hevea in view of this?

Unless there's a browser interoperability problem I think you can close these issues. According to [this](https://drafts.csswg.org/css2/syndata.html#value-def-identifier) no invalid identifiers are being generated and it is quite easy to escape these `:`.

**Comment by Chris00** [comment]
@aantron @sanette I have not seen any browser issue (Firefox, Chromium).  Please let me know if it is different for any of you.

**Comment by sanette** [comment]
as far as I'm concerned, this is fine.

(Anyway I prefer not to use escaping to ensure I'm on a robust side, so I'm using the alternative ways mentioned above.)

**Comment by Chris00** [comment]
IMHO, these ways should be mentioned in lambdasoup documentation.  That said, escaping is part of the CSS spec and should therefore be considered robust.  Maybe lambdasoup could even provide an `escape` function.
